### DISCUSSION TITLE
LAN sensors for battery_charge_total, etc

### DISCUSSION THREAD

**Discussion by canton7** [nathanmarlor/foxess_modbus#166] [General]
Hi @nathanmarlor,

When going through the H3 registers, I found that there are actually holding registers for:

 - battery_charge_total
 - battery_discharge_total
 - feed_in_energy_total
 - grid_consumption_energy_total

The problem is, we've already got integration sensors for these. If we switch the sensor type, I'm not sure what will happen: users will either see a jump, or they'll lose the history.

What's worse, is that these sensors are used by the energy dashboard. So if they jump, the user's energy stats will be all over the place.

I guess another alternative is to do a config migration, and add a key saying to keep legacy sensor types. Then people who have the integration sensors will keep them, and new users will get the new, proper sensors.

Any preferences on what we do here?

**Comment by canton7** [comment]
On the other hand, the longer we leave it, the larger the impact will be on users...

**Comment by screenagerbe** [comment]
They can save their statistics, but you have to be very careful when trying to. and it is only possible since HA 2023.4
I managed to do it in my TST environment, but I screwed it partly up in PROD. 
you have to disable the integration which uses the current entities, then you rename this entities with the name from the new sensor, next you install/update the integration

![image](https://user-images.githubusercontent.com/82099983/233153523-7e8f3323-89fb-4381-9259-d3e228edeb26.png)
![image](https://user-images.githubusercontent.com/82099983/233153753-df3db5b7-621e-4a8e-a0c9-2e87478b1802.png)

Guess you can see where it went wrong 🙄
But then again I have grafana running so I can select the proper entities to visualise them nicely

**Reply by canton7** [reply]
This isn't about renaming entities. It's about changing an entity from an intigration sensor (where HA is summing energy to get power) to a normal sensor which is fetching power directly from the inverter.

**Reply by screenagerbe** [reply]
yeah I get that.
But because you're renaming them (and the original integration isn't available anymore) you have the opportunity to re-claim the statistics, and that's the ultimate goal.
Anyway, if you really want to re-claim the stats, it is possible, but it is a bit of hassle.

**Reply by canton7** [reply]
When I say integration sensor, I mean Reimann integral sensor. Not a sensor from a different integration.

**Reply by canton7** [reply]
Eg https://github.com/nathanmarlor/foxess_modbus/blob/main/custom_components/foxess_modbus/entities/xx1_lan_entity_descriptions.py#L155

**Comment by canton7** [comment]
The more I think about it, the most I like the idea of "entity set versions". Each entity description says what entity set versions it should be present in. When the user installs the integration, we record whatever the latest entity set version is at that point. Thereafter, we only show entities which should be present in that version.

That lets us do non-breaking changes for users, but means they're opted out of breaking changes (renames, changed types, etc).

We could have a button in config to bring them into the latest version... 

That said, I don't know how many more of these we'll get. Hopefully we're getting a handle on what the inverters support!
